programs/mango-v4/src/state/oracle.rs

```diff
@@ -357,50 +412,55 @@ pub fn oracle_state_unchecked(
             }
         }
         OracleType::OrcaCLMM => {
-            let usd_state = usdc_state_unchecked(usd_feed_opt)?;
-
-            let whirlpool = Whirlpool::try_deserialize(&mut &data[..]).unwrap();
-
+            let whirlpool = load_whirlpool_state(oracle_info)?;
             let sqrt_price = U64F64::from_bits(whirlpool.sqrt_price);
 
-            let clmm_price = if whirlpool.token_mint_a == usdc_mint_mainnet::ID {
-                // inverted
-                let decimals = QUOTE_DECIMALS - (base_decimals as i8); // tokenMintB - tokenMintA
-                let decimal_adj = power_of_ten(decimals);
-                let inv_price = I80F48::from_num(sqrt_price * sqrt_price) * decimal_adj;
-                Ok(I80F48::ONE.checked_div(inv_price).unwrap())
-            } else if whirlpool.token_mint_b == usdc_mint_mainnet::ID {
-                let decimals = (base_decimals as i8) - QUOTE_DECIMALS; // tokenMintA - tokenMintB
-                let decimal_adj = power_of_ten(decimals);
-                Ok(I80F48::from_num(sqrt_price * sqrt_price) * decimal_adj)
+            let inverted = whirlpool.token_mint_a == usdc_mint_mainnet::ID
+                || (whirlpool.token_mint_a == sol_mint_mainnet::ID
+                    && whirlpool.token_mint_b != usdc_mint_mainnet::ID);
+            let quote_state = if inverted {
+                quote_state_unchecked(acc_infos, &whirlpool.token_mint_a)?
+            } else {
+                quote_state_unchecked(acc_infos, &whirlpool.token_mint_b)?
+            };
+
+            let clmm_price = if inverted {
+                let inverted_price = I80F48::from_num(sqrt_price * sqrt_price);
+                I80F48::ONE.checked_div(inverted_price).unwrap()
```

Comment:
1. no need to use checked_div, checked math is enabled for all I80F48 ops
2. we likely want to use f64 division anyway because it's way cheaper (2000 CU vs 200 CU)

For the latter, the cheapest path would be to convert sqrt_price to f64 very early (costs 100 CU), then do the squaring (200 CU cheaper as f64), do the division (1800 CU cheaper as f64) and then convert back to I80F48 (costs 200 CU). It's sad that this is only cheaper when the division is needed, so need two different codepaths?